crates/noirc_evaluator/src/ssa/acir_gen/internal_var_cache.rs

```diff
@@ -31,8 +37,24 @@ impl InternalVarCache {
 
         let mut var = match ctx.try_get_node(id)? {
             NodeObject::Const(c) => {
+                // use the InternalVar from constants if exists
                 let field_value = FieldElement::from_be_bytes_reduce(&c.value.to_bytes_be());
-                InternalVar::from_constant(field_value)
+                let mut result = InternalVar::from_constant(field_value);
+                if let Some(c_var) = self.constants.get(&field_value) {
+                    result = c_var.clone();
+                }
+                //use witness from other nodes if exists
+                if let Some(constant_ids) = ctx.constants.get(&field_value) {
+                    for c_id in constant_ids {
+                        if let Some(i_var) = self.inner.get(c_id) {
+                            if i_var.cached_witness().is_some() {
+                                result.set_witness(*i_var.cached_witness());
+                                break;
+                            }
+                        }
+                    }
+                }
```

Comment:
```suggestion
                if let Some(constant_ids) = ctx.constants.get(&field_value) {
                    let maybe_cached_witness = constant_ids
                        .iter()
                        .find_map(|c_id| self.inner.get(c_id).map(|i_var| i_var.cached_witness()));

                    if let Some(cached_witness) = maybe_cached_witness {
                        result.set_witness(*cached_witness)
                    }
                }
```

This makes the logic clearer and avoids some rightward shift.

Reply:
Calling 5 functions in one line and returning an option of option does not make it cleaner for me.

Reply:
Hmm I think the fact that we have an Option of Option is already quite strange -- I tend to lean towards reducing rightwards shift as much as possible, so I'd lean towards Tom here after looking at this code snippet again.

Here is some pseudo-code that shows something in the middle:

```rust
let constant_ids = match ctx.constants.get(&field_value) {
   Some(constant_ids) => constant_ids,
   None => return, // early return
};

for c_id in constant_ids {
   let i_var = match self.inner.get(c_id) {
      Some(i_var) => i_var,
      None => return,
   };
   if let Some(cached_witness) = i_var.cached_witness() {
      result.set_witness(cached_witness);
      break;
   }
}


```

Reply:
Thank you for both of your comments. While we don't agree on what is the best style because of different sensibilities, I think I came up, using your both inputs, with a middle-ground solution that has no indentation and is still readable for me.

Reply:
I like the separation of search + insertion in the new implementation :+1: